_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+    of just `bool`.
+    We are leaning towards the latter, but it needs more consideration.
+1. `++` and `--` are very convenient, and we want to have them eventually.
+    They just need to be handled in a special way, and there is more than way to do so.
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
+1. We are considering adding a C++-like
+    ["spaceship operator"](https://en.wikipedia.org/wiki/Three-way_comparison#High-level_languages),
+    i.e. `<=>`.
+    When defined, the compiler could use it to automatically generate all the other comparison operators.
+
+On the other hand, it is unlikely that we will allow redefining some other operators:
+
+1. `&&` and `||` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    The additional complication is their short-circuiting behavior that may or may not be
+    appropriate, depending on the semantics of a particular UDVT.
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
+
+#### Operators on Reference Types
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
+The main thing holding us back are the planned changes to reference types, in particular the fact
+that [data location](https://docs.soliditylang.org/en/latest/types.html#data-location) is going to
+become a part of the type itself.
+ To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
+
+We want to eventually support operators on other reference types, but it is likely that this usage will
+require naming them, using some kind of yet-to-be-introduced user-defined reference type feature similar
+to UDVTs, or at least defining a type alias.
+
+#### Operators on Mixed Types
+Inability to define operators with arguments of different types is somewhat constraining and lifting this
+limitation is also a part of our vision for UDVTs.
+There are, however, multiple ways to achieve this, and explicitly defining an operator for every type combination
+is only one of them.
+Other possibilities involve generics specialized with traits or having implicit conversions.
+
+This part will unfortunately have to wait until we are done with the changes to the type system that are on our current roadmap.
+Doing them before would only result in something that would limit our design choices, and would have to be discarded
+in a breaking release.
```

Comment:
I'm not sure this is true :-). Mixed type operators are a significant problem for type inference, so I'm not sure we'll ever want to have them.

I think it would be better to be very conservative on this and merely state that the first version we're releasing is a minimal viable implementation that will definitely not cause problems if considered as part of a generic type system or trait system with type inference without making any promises or talking about mixed operators as part of a vision, which they aren't necessarily.

Reply:
How about this?

```
Inability to apply operators to values of different types is somewhat constraining and lifting this
limitation is something we would like to address eventually, though it is not yet clear in what form.
There are multiple ways to achieve this.
Explicitly defining an operator for every type combination is just one of them, and not necessarily the most elegant.
For the time being, we can only say that the feature we are releasing is a minimal viable implementation
that will definitely not cause problems if considered as part of a generic type system or trait system
with type inference.
```

But I'm not sure I understand what you mean. Are you saying that ending up without operators that work on more than one type is an real possibility? I thought we wanted to allow that, one way or another. I.e. the classic example of adding `Timestamp` and `Duration` that @chriseth often brings up. That has to be possible eventually, doesn't it?

Reply:
The whole section about the future of operators removed.

Reply:
I swayed @chriseth away from it by now :-). Please don't promise mixed operators in any way, we'll likely never have them and it will be hard to explain to people why already, since it's a type-theoretic problem of the ambiguity of type inference that's at stake here that's not so easy to convey.